### PR TITLE
JPERF-522: Added recallByTag method JqlMemory (default) and implementation

### PR DESCRIPTION
**Pull request by drspartez**
Initial discussion was here https://github.com/atlassian/jira-actions/pull/16
This is not invasive change that provides ability to create SearchJql actions  with user selected JQLS

### CODE REVIEW

File: src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import org.mockito.stubbing.Answer
+import org.openqa.selenium.WebDriver
+
+class JqlMemoryTest {
+
+    @Test
+    fun allBuiltInQueryTypesAvailable() {
+        val memory: JqlMemory = AdaptiveJqlMemory(SeededRandom());
+        val webdriverMock = Mockito.mock(WebDriver::class.java)
+        val issuePageSpy = Mockito.spy(IssuePage(webdriverMock))
```
Comment (dagguh) [inline src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt]: This is unfortunately necessary if we want to mock `IssuePage`.
Ideally, we would turn `IssuePage` into an interface and use a mock implementation here, avoiding the huge runtime, dependency, readability and diagnosability cost of Mockito.

Changing a class into an interface is binary-incompatible (`invokeinterface` bytecode instruction). Well be able to clean this up only in a major release.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/adaptive/AdaptiveJqlMemory.kt [hunk 2]
```diff
@@ -14,42 +15,55 @@ class AdaptiveJqlMemory(
...
+
 }
 
 data class BakedJql(
     val jqlPrescription: JqlPrescription,
-    val jql: String?
-)
\ No newline at end of file
+    val jql: String?,
+    val tag: String?
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/adaptive/AdaptiveJqlMemory.kt]: This breaks `copy` calls, because it's a `data class`. We can assume nobody calls `copy`, but we've been wrong like this before.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt [hunk 3]
```diff
@@ -3,5 +3,10 @@ package com.atlassian.performance.tools.jiraactions.api.memories
 import com.atlassian.performance.tools.jiraactions.api.page.IssuePage
 
 interface JqlMemory : Memory<String> {
+
     fun observe(issuePage: IssuePage)
+
+    fun recallByTag(filter: (String) -> Boolean): String? {
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: This requires `kotlin.jvm.functions.Function1`.
We can make it more portable if we switch to `java.util.function.Predicate`.

Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: This uses the Kotlin default implementation, which is not consumable by Java implementors, which breaks them, unless we use `@JvmDefault`.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt [hunk 4]
```diff
@@ -3,5 +3,10 @@ package com.atlassian.performance.tools.jiraactions.api.memories
 import com.atlassian.performance.tools.jiraactions.api.page.IssuePage
 
 interface JqlMemory : Memory<String> {
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: It seems like we want to treat JQLs not only as `String`s, but something more. We want to categorize them, tag them. We want a `Memory<Jql>`. This is why [primitive obsession](http://wiki.c2.com/?PrimitiveObsession) is a ticking time bomb. It strips us from extension points.

This is why the primitive obsession in `recallByTag` will also create problems in the future.

I propose to add a brand new interface, use domain classes (like `class Jql(val query: String, type: JqlType)`) and leave `JqlMemory` intact. The `AdaptiveJqlMemory` can implement both old and new ones at the same time.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt (hunk 3 above)
Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: It is done the same fashion as IssueMemory recall is.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt (hunk 4 above)
Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: I considered this, but I'm not doing refactoring this time.

File: src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt (hunk 1 above)
Comment (drspartez) [reply src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt]: I do not see reason to write more code or do refactoring when my objective different. I see mockito is widely used in atlassian/spartez projets. Because of mockito this test case is clean and neat. 
I dont't get why we would be able to clean this in major release - this is the same module.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt (hunk 3 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: @drspartez IssueMemory is already part of an API. It's hard to fix it. I see no point in repeating the mistake here.

Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: Ok. I didn't know t is mistake.

Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: Ok

File: src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt (hunk 1 above)
Comment (wyrzyk) [reply src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt]: I think runtime and dependency costs doesn't need explanation. I want to add something about readability and diagnosability:

- A proper mock could be reused across multiple tests. We have just two tests here, but we need to repeat the blocks of code that creates the mock.
- It can introduce bugs in the future. Without Mockito, when changing mocked objects, you need to adjust your mocks. You need to make a conscious decision, how the mock should behave.  Mockito can transparently adapt for most cases. In the end, you may end up testing something else than intended (we hade examples of the cases like that internally). 
- Like every reflection-based tool, it may fool IDE. Searching for usages of a method may not work well (I'm not sure if it's still an issue in IJ 2019, but it's an open-source project and developers can use any IDE).
- If you put complex logic into your mock, then it's hard to debug (reflection again). It's ok if you return values, but by introducing the library on a classpath, you encourage developers to use it for every case.
- We have multiple reasons to write tests. To protect against regressions. To document how to use an API. To reproduce the issue. Another reason is to verify if the API is useable. Mockito makes mocking unmockable objects easy. It means we'll stop to see problematic places and we'll never improve it. Problems with testing are very similar to issues of the code and reuse of it. We should treat our tests as the first user of an API.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt [hunk 5]
```diff
@@ -1,7 +1,14 @@
...
 import com.atlassian.performance.tools.jiraactions.api.page.IssuePage
+import java.util.function.Predicate
 
 interface JqlMemory : Memory<String> {
+
     fun observe(issuePage: IssuePage)
+
+    @JvmDefault
+    fun recallByTag(filter: Predicate<String>): String? {
+        return null;
```
Comment (lipinskipawel) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: the semicolon at the of the `return null`

Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: I'm also concerned about returning `null` as a default implementation. It will make the change source compatible, but I see no value besides that. It'll probably hide the problem until it appears in a different place (no results for the action at the end of the test).

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt (hunk 4 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: @drspartez When making a tradeoff like this, it's a good practice to describe it in a commit's description. It adds a couple of extra benefits:
- You need to document the decision, so you need to make a logically sound block of text. For some persons (including me) it helps to sort things out. Sometimes it may affect a decision.
- Other people can review your decision. It'll help them to understand it and may help to find wholes in the reasoning. It may also affect the decision.
- If a change is confusing for the current readers, then it'll be confusing for future readers too. Commits are easier to access than PR.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt (hunk 5 above)
Comment (lipinskipawel) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: Maybe this method should be documented e.g which strings are allowed? Can we somehow validate them? Should we change the name of the method into recallByPredicate? We can throw an Exception as the default implementation.

File: src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt (hunk 1 above)
Comment (drspartez) [reply src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt]: > * A proper mock could be reused across multiple tests. We have just two tests here, but we need to repeat the blocks of code that creates the mock.

I used to create mock to cover some specific code path and i think it is right.
You're right about redundant mock preparation code.

>* It can introduce bugs in the future. Without Mockito, when changing mocked objects, you need to adjust your mocks. You need to make a conscious decision, how the mock should behave. Mockito can transparently adapt for most cases. In the end, you may end up testing something else than intended (we had examples of the cases like that internally).

Mockito requires using explicit API calls to prepare mocked instance so it is cean how API is used and mocked. Involving mockito or not has nothing to do with test scope/goal/intention.

> * Like every reflection-based tool, it may fool IDE. Searching for usages of a method may not work well (I'm not sure if it's still an issue in IJ 2019, but it's an open-source project and developers can use any IDE).

There are explicit API calls

> * If you put complex logic into your mock, then it's hard to debug (reflection again). It's ok if you return values, but by introducing the library on a classpath, you encourage developers to use it for every case.

There is no complex logic in here

> * We have multiple reasons to write tests. To protect against regressions. To document how to use an API. To reproduce the issue. Another reason is to verify if the API is useable. Mockito makes mocking unmockable objects easy. It means we'll stop to see problematic places and we'll never improve it. Problems with testing are very similar to issues of the code and reuse of it. We should treat our tests as the first user of an API.

This is separate concern (unmockable objects and its improvement) and mockito does not caused the problem and does not solves it (in terms of design). When you wrote dozen lines of code without mockito and it is working for you, then this also causes problematic places are not visible.
Mockito helps to show how api can be used . Ido not see this is broken in area of test concern

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt (hunk 5 above)
Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: Default recall also returns null when there is no result i think it is better than exception.
In my opinion name is ok because it provides more context. Everybody see that predicate is there

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt (hunk 4 above)
Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: This is something new to me in scope of commit message. I was used to deal with project/design decisions on issue level. But here it is brainstormed more on PR. So i don't get why commit message appears here.

File: src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt (hunk 1 above)
Comment (wyrzyk) [reply src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt]: >I used to create mock to cover some specific code path and i think it is right.
You're right about redundant mock preparation code.

It's ok. It just violates the DRY principle. For some cases, redundancy can be better than a broken abstraction. I would not put much weight to that.

> Mockito requires using explicit API calls to prepare mocked instance so it is cean how API is used and mocked. Involving mockito or not has nothing to do with test scope/goal/intention.

This is an internal example of the problem I've mentioned https://hello.atlassian.net/wiki/spaces/~kcichy/blog/2018/10/12/321652230/Do+you+trust+your+mock#Mocks-go-wrong-over-time. It's clear how to use and mock objects. It'll work for most cases at the time when the test is written. But Mockito helps mocks to deteriorate unnoticeably.


> There is no complex logic in here


I'm not concerned with complex logic here. I'm concerned about giving a tool (on a classpath) that can be misused in the future.

> This is separate concern (unmockable objects and its improvement) and mockito does not caused the problem and does not solves it (in terms of design). When you wrote dozen lines of code without mockito and it is working for you, then this also causes problematic places are not visible.
Mockito helps to show how api can be used . Ido not see this is broken in area of test concern

 
Mockito makes dealing with problematic API easy. Writing dozens of lines of code is painfull, so it reminds about the problem. It's still easier to write the mocks than to fix the API, but the problem lays in a different place. The problem may be invisible (it's easy to mock with Mockito) for a new code. We may introduce a new broken API that's hard to mock because we have a tool with super-powers.

Comment (drspartez) [reply src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt]: @wyrzyk 
I do not see that reason for the problem with test mentioned in the blogpost of Kamil was mockito. If the test was written with manual mocks results would be the same.
Primary reason that the test didn't catch the problem is not implemented coverage.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt (hunk 4 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: > So i don't get why commit message appears here.

It appears here to indicate that we could avoid confusion like in this thread in the future if we document decisions. I also added reasons why it's a good practice to document it close to the source. 


decision documented in a commit description >> decision documented in the issue >> decision documented in the PR >> decision not documented at all.

For some cases, code comments would be best for documenting a decision. I'd leave it out of the scope of the discussion, as it highly depends on the context.

Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: @wyrzyk Would You like to propose neat pattern for such description relevant for commit message ?

Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: I'm not the best person in case of writing human-readable text, but I'll try:
```
JPERF-522: Add support for filtering by JQL type in JqlMemory

JqlMemory operates on Strings instead of JQL objects. It makes evolving the API
hard. I decided to keep using String for this change to avoid API refactoring costs.
```

Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt]: Cool. Thanks

**Comment by mcmurphy2000** [comment]
Guys we need to move with this, I think there was enough discussion, some things were resolved, there some disagreement but it's blocking our current sprint. We need to move on.

File: src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt (hunk 1 above)
Comment (drspartez) [reply src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt]: All test stuff is reverted. Not because I do agree with this review, but because I'm forced by time.

**Comment by drspartez** [comment]
Test to this code will appear with next PR related to JQL action split.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/adaptive/AdaptiveJqlMemory.kt [hunk 6]
```diff
@@ -5,48 +5,61 @@ import com.atlassian.performance.tools.jiraactions.api.memories.JqlMemory
...
     }
+
+    override fun recallByTag(filter: Predicate<String>): String? {
+        return random.pick(jqls.filter {it.tag != null && filter.test(it.tag.toString()) }.toList())?.baked?.jql
+    }
+
+    private class TaggedBakedJql(val baked: BakedJql, val tag: String? = null)
 }
 
 data class BakedJql(
```
Comment (lipinskipawel) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/adaptive/AdaptiveJqlMemory.kt]: Since we are doing a minor change, perhaps we can deprecate this class? 
Why this class should stay public?

Comment (drspartez) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/adaptive/AdaptiveJqlMemory.kt]: I see this class as candidate to deprecate when we introduce some String with JQL representation class - it would be exact replacement.

File: src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt (hunk 1 above)
Comment (dagguh) [reply src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt]: @drspartez I think there's been a misunderstanding. I agreed that Mockito is necessary in this case and that we should go for it. Having a test with Mockito is better than not having a test at all.

I meant to inform that we wouldn't have to use Mockito if `IssuePage` was better designed. We can improve its design only in a major release. Then we should remove Mockito.

Being forced to resort to reflection hacks like Mockito proves how hard it is to customize `IssuePage`. Currently users have to settle on copy-pasting implementations.

TL;DR: Mockito is bad, but this time it was necessary.
